inst/include/dust/gpu/rng.hpp

```diff
@@ -0,0 +1,144 @@
+#ifndef DUST_GPU_RNG_HPP
+#define DUST_GPU_RNG_HPP
+
+#include <dust/gpu/xoshiro.hpp>
+#include <dust/gpu/distr/binomial.hpp>
+#include <dust/gpu/distr/normal.hpp>
+#include <dust/gpu/distr/poisson.hpp>
+#include <dust/gpu/distr/uniform.hpp>
+
+namespace dust {
+
+struct RNGptr {
+  uint64_t* state_ptr;
+  unsigned int state_stride;
+  unsigned int particle_stride;
+};
+
+// Read state from global memory
+template <typename T>
+__device__
+rng_state_t<T> loadRNG(RNGptr& rng_state, int p_idx) {
+  rng_state_t<T> state;
+  for (int i = 0; i < XOSHIRO_WIDTH; i++) {
+    state.s[i] = rng_state.state_ptr[p_idx * rng_state.particle_stride +
+                                      i * rng_state.state_stride];
+  }
+  return state;
+}
+
+// Write state into global memory
+template <typename T>
+__device__
+void putRNG(rng_state_t<T>& rng, RNGptr& rng_state, int p_idx) {
+  for (int i = 0; i < XOSHIRO_WIDTH; i++) {
+    rng_state.state_ptr[p_idx * rng_state.particle_stride +
+                        i * rng_state.state_stride] = rng.s[i];
+  }
+}
+
+template <typename real_t>
+class pRNG { // # nocov
+public:
+  pRNG(const size_t n, const uint64_t seed) {
+    dust::Xoshiro rng(seed);
+    for (int i = 0; i < n; i++) {
+      _rngs.push_back(rng);
+      rng.jump();
+    }
+
+    CUDA_CALL(cudaMalloc((void** )&_d_rng_state.state_ptr,
+                          n * XOSHIRO_WIDTH * sizeof(uint64_t)));
+    // Set to be interleaved
+    _d_rng_state.state_stride = n;
+    _d_rng_state.particle_stride = 1;
```

Comment:
Add `assert(_d_rng_state.state_stride * _d_rng_state.particle_stride == n)` or something better if you don't like asserts

Reply:
Added in 57477d4